packages/expo-image-manipulator/src/actions/CropAction.web.ts

```diff
@@ -0,0 +1,35 @@
+import { CodedError } from 'expo-modules-core';
+
+import { ActionCrop } from '../ImageManipulator.types';
+import { getContext } from '../utils/getContext.web';
+
+export default (canvas: HTMLCanvasElement, options: ActionCrop['crop']) => {
+  // ensure values are defined.
+  let { originX = 0, originY = 0, width = 0, height = 0 } = options;
+  const clamp = (value, max) => Math.max(0, Math.min(max, value));
+  // lock within bounds.
+  width = clamp(width, canvas.width);
+  height = clamp(height, canvas.height);
+  originX = clamp(originX, canvas.width);
+  originY = clamp(originY, canvas.height);
+
+  // lock sum of crop.
+  width = Math.min(originX + width, canvas.width) - originX;
+  height = Math.min(originY + height, canvas.height) - originY;
+
+  if (width === 0 || height === 0) {
+    throw new CodedError(
+      'ERR_IMAGE_MANIPULATOR_CROP',
+      'Crop size must be greater than 0: ' + JSON.stringify(options, null, 2)
+    );
+  }
+
+  const result = document.createElement('canvas');
```

Comment:
I'm curious about creating this separate `canvas` and obtaining the context out of it.
How expensive is that in terms of resources? 🤔
Is it possible to reuse the canvas and possibly the context? Just wondering 🤔
Correct me if I'm wrong, but after the action is performed on the canvas (using some context) we can safely assume that this context (and it's configuration) and canvas (and it's configuration, apart from storing the result) are not longer needed.
Don't get me wrong - I'm just curios if we can get even better in terms of performing the actions ✊

Reply:
> I'm curious about creating this separate canvas and obtaining the context out of it.
> How expensive is that in terms of resources? 🤔

I don't think there is particularly an overhead in creating a new canvas and getting the context.

I quickly put together a benchmark of creating a 1024x1024 canvas and drawing another canvas on it vs creating the canvas and getting the context without any drawing.

[Benchmark canvas draw vs no draw](https://jsfiddle.net/miken92/s7frwu50/) - view browser console or open it bottom right.

Different browsers do seem to vary quite a lot (Safari seems very good at draw2 - >500,000 per second).

MBP 2017 (Chrome):
"draw1 x 1,651 ops/sec ±20.59% (36 runs sampled)"
"draw2 x 52,357 ops/sec ±1.13% (85 runs sampled)"
"Fastest is draw2"

---

> Is it possible to reuse the canvas and possibly the context? Just wondering 🤔

I don't think so, the issue is we need to set our transformation options before drawing and as demonstrated it's pretty fast drawing canvas to canvas (I assume because it can all stay on the GPU).

For instance a rotate is:
- Create result canvas/context
- Setup transformation (translate and rotate)
- Draw the previous canvas

I think this is possibly why toDataURL operations are so slow if each time it is called it needs to actually create an image rather than whatever kind of of representation its kept as in memory.

I actually had a quick play around with [fabricjs/fabric.js](https://github.com/fabricjs/fabric.js) to stack all the actions/operations together, I'm not sure how it works internally but it was way slower than the simple canvas operations.

```
...
} else if ('flip' in action) {
  if (action.flip === FlipType.Horizontal) {
    options.flipX = !(options.flipX ?? false);
  } else if (action.flip === FlipType.Vertical) {
    options.flipY = !(options.flipY ?? false);
  }
  // return flip(canvas, action.flip);
} else if ('rotate' in action) {
  options.angle = (options.angle ?? 0) + action.rotate;
  // return rotate(canvas, action.rotate);
...
```

---

> Correct me if I'm wrong, but after the action is performed on the canvas (using some context) we can safely assume that this context (and it's configuration) and canvas (and it's configuration, apart from storing the result) are not longer needed.

I believe you are correct there will be a period of time where multiple canvases/context exist with drawings on them taking up resources, but as far as I know as long as they fit they should be in GPU memory.
That is until the garbage collector disposes of them.
But that's no different to the previous implementation which was creating a canvas/context per action but returning a data URL.

---

I've had a look at profiling the performance snack with my changes (ran on desktop pc - takes about 2.5 seconds when not profiling)

![image](https://user-images.githubusercontent.com/9887246/130763096-209761a9-4143-4647-9732-a1fd3273f163.png)

As you can see the actual drawing operations are extremely quick (with the slower functions being the JavaScript based resize and toDataURL).

Some future possible optimisations I'm thinking of exploring

- toDataURL will lock up the UI, I think this can be avoided using `toBlob` and `URL.createObjectURL` I don't think it'll be any quicker but a nicer UX although I believe developers would need to know to dispose the URL when its no longer needed and it'd be a breaking change.
- `resampleSingle` the [author](https://stackoverflow.com/a/18320662/4047926) wrote a newer version that uses web workers that may be worth investigating